### PR TITLE
[MLA-1634] Add ObservationSpec and update ISensor interfaces

### PR DESCRIPTION
**Pull request by chriselion**
### Proposed change(s)
* Add an InplaceArray<T> implementation for "blittable" arrays with a max size of 4. This is used for ObservationSpec shapes and dimension properties.
* Add an ObservationSpec struct and helper methods.
* Change ISensor interface to return ObservationSpec instead of just a shape array.
* Remove IDimensionPropertiesSensor and ITypedSensor interfaces, use information from ObservationSpec instead.

Note that this does *not* introduce CompressionSpec or remove ISparseChannelSensor yet; this was getting big, so I'm saving that for a subsequent PR.

I haven't updated the docs or migration guide yet, will do that soon. Test coverage is done though.

### Follow-Ups in subsequent PRs
* Convert SensorComponent too
* Introduce CompressionSpce and remove ISparseChannelSensor

### Useful links (Github issues, JIRA tickets, ML-Agents forum threads etc.)
https://jira.unity3d.com/browse/MLA-1634
Alternative to https://github.com/Unity-Technologies/ml-agents/pull/5076
(Unity internal) [design proposal](https://docs.google.com/document/d/1DBOIv7NdKTQn3fMNerKvZUpfilcWfcbBG25q1hNZcTc/edit#) 

### Types of change(s)
- [x] New feature
- [x] Code refactor
- [x] Breaking change

### Checklist
- [x] Added tests that prove my fix is effective or that my feature works
- [x] Updated the [changelog](https://github.com/Unity-Technologies/ml-agents/blob/main/com.unity.ml-agents/CHANGELOG.md) (if applicable)
- [ ] Updated the [documentation](https://github.com/Unity-Technologies/ml-agents/tree/main/docs) (if applicable)
- [x] Updated the [migration guide](https://github.com/Unity-Technologies/ml-agents/blob/main/docs/Migrating.md) (if applicable)

### CODE REVIEW

File: DevProject/ProjectSettings/Packages/com.unity.testtools.codecoverage/Settings.json [hunk 1]
```diff
@@ -2,6 +2,22 @@
     "m_Name": "Settings",
     "m_Path": "ProjectSettings/Packages/com.unity.testtools.codecoverage/Settings.json",
     "m_Dictionary": {
-        "m_DictionaryValues": []
+        "m_DictionaryValues": [
```
Comment (chriselion) [inline DevProject/ProjectSettings/Packages/com.unity.testtools.codecoverage/Settings.json:5]: Not sure what to do with this. It got added in https://github.com/Unity-Technologies/ml-agents/commit/7718dfe11da14b5c21dd5c32e4ce2a3a03ed8ca4#diff-41dd07c8a0392e1f6c2a7ba1277f3426cdc92aa8258e563912c39db66800ade0 but wasn't computing coverage on the assemblies we care about.

File: com.unity.ml-agents.extensions/Runtime/Sensors/GridSensor.cs [hunk 2]
```diff
@@ -475,14 +475,6 @@ public void ClearPerceptionBuffer()
             }
         }
 
-        /// <summary>Gets the shape of the grid observation</summary>
-        /// <returns>integer array shape of the grid observation</returns>
-        public int[] GetFloatObservationShape()
```
Comment (chriselion) [inline com.unity.ml-agents.extensions/Runtime/Sensors/GridSensor.cs:480]: I think this was unused.

File: com.unity.ml-agents.extensions/Runtime/Sensors/GridSensor.cs [hunk 3]
```diff
@@ -914,10 +906,22 @@ void ISensor.Update()
...
+            }
+            return m_ObservationSpec;
+        }
+
+        /// <inheritdoc/>
         public override int[] GetObservationShape()
         {
-            m_Shape = new[] { GridNumSideX, GridNumSideZ, ObservationPerCell };
-            return m_Shape;
+            var shape = m_ObservationSpec.Shape;
```
Comment (chriselion) [inline com.unity.ml-agents.extensions/Runtime/Sensors/GridSensor.cs:923]: This is the SensorComponent override, not sure if we should change that to return a spec too.

File: com.unity.ml-agents/Runtime/Actuators/IDiscreteActionMask.cs [hunk 4]
```diff
@@ -10,6 +10,7 @@ public interface IDiscreteActionMask
         /// <summary>
         /// Set whether or not the action index for the given branch is allowed.
         /// </summary>
+        /// <remarks>
```
Comment (chriselion) [inline com.unity.ml-agents/Runtime/Actuators/IDiscreteActionMask.cs:13]: Not from here, but was broken in staging.

File: com.unity.ml-agents/Runtime/InplaceArray.cs [hunk 5]
```diff
@@ -0,0 +1,250 @@
...
+using System.Collections.Generic;
+
+namespace Unity.MLAgents
+{
+    /// <summary>
+    /// An array-like object that stores up to four elements.
+    /// This is a value type that does not allocate any additional memory.
+    /// </summary>
+    /// <remarks>
+    /// This does not implement any interfaces such as IList, in order to avoid any accidental boxing allocations.
```
Comment (chriselion) [inline com.unity.ml-agents/Runtime/InplaceArray.cs:11]: Small tradeoff between ergonomics and efficiency. It's small either way, so I'm willing to change this.

File: com.unity.ml-agents/Runtime/InplaceArray.cs [hunk 6]
```diff
@@ -0,0 +1,250 @@
...
+            }
+        }
+
+        /// <summary>
+        /// Check that the arrays have the same length and have all equal values.
+        /// </summary>
+        /// <param name="lhs"></param>
+        /// <param name="rhs"></param>
+        /// <returns>Whether the arrays are equivalent.</returns>
+        public static bool operator==(InplaceArray<T> lhs, InplaceArray<T> rhs)
```
Comment (chriselion) [inline com.unity.ml-agents/Runtime/InplaceArray.cs:198]: What's the recommended way to do operator== and Equals() for a value type? Should the implementation be in Equals() instead?

File: com.unity.ml-agents/Runtime/InplaceArray.cs [hunk 7]
```diff
@@ -0,0 +1,250 @@
...
+            return this == other;
+        }
+
+        /// <summary>
+        /// Get a hashcode for the array.
+        /// </summary>
+        /// <returns></returns>
+        public override int GetHashCode()
+        {
+            return Tuple.Create(m_Elem0, m_Elem1, m_Elem2, m_Elem3, Length).GetHashCode();
```
Comment (chriselion) [inline com.unity.ml-agents/Runtime/InplaceArray.cs:247]: Not currently used anywhere, but I had to add this if I overrode Equals().

File: com.unity.ml-agents/Runtime/Sensors/ISensor.cs [hunk 8]
```diff
@@ -16,18 +16,72 @@ public enum SensorCompressionType
...
+        /// Means that there can be a variable number of observations in this dimension.
+        /// The observations are unordered.
+        /// </summary>
+        VariableSize = 4,
+    }
+
+    /// <summary>
+    /// The ObservationType enum of the Sensor.
+    /// </summary>
+    public enum ObservationType
```
Comment (chriselion) [inline com.unity.ml-agents/Runtime/Sensors/ISensor.cs:51]: These were made public, but didn't make any other changes to them. We'll remove Reward and Message, and rename Goal later.

File: com.unity.ml-agents/Runtime/Sensors/ObservationSpec.cs [hunk 9]
```diff
@@ -0,0 +1,108 @@
...
+    /// should be used for training.
+    /// </summary>
+    public struct ObservationSpec
+    {
+        /// <summary>
+        /// The size of the observations that will be generated.
+        /// For example, a sensor that observes the velocity of a rigid body (in 3D) would use [3].
+        /// A sensor that returns an RGB image would use [Height, Width, 3].
+        /// </summary>
+        public InplaceArray<int> Shape;
```
Comment (chriselion) [inline com.unity.ml-agents/Runtime/Sensors/ObservationSpec.cs:17]: Open to discussion on the visibility for these fields, and also the constructor. I think the static helpers (Vector, VariableLength, Visual) are good for all the purposes that we currently support in the trainer, and a user could potentially implement their own trainer that supported different combos of properties (e.g. convolutional 1D or 3D).

File: com.unity.ml-agents/Runtime/Sensors/ObservationSpec.cs [hunk 10]
```diff
@@ -0,0 +1,108 @@
...
+        /// <summary>
+        /// The type of the observation, e.g. whether they are generic or
+        /// help determine the goal for the Agent.
+        /// </summary>
+        public ObservationType ObservationType;
+
+        /// <summary>
+        /// The number of dimensions of the observation.
+        /// </summary>
+        public int NumDimensions
```
Comment (chriselion) [inline com.unity.ml-agents/Runtime/Sensors/ObservationSpec.cs:39]: Should we call this Rank instead?

File: com.unity.ml-agents/Runtime/Sensors/SensorShapeValidator.cs [hunk 11]
```diff
@@ -34,13 +34,14 @@ public void ValidateSensors(List<ISensor> sensors)
...
                 for (var i = 0; i < Mathf.Min(m_SensorShapes.Count, sensors.Count); i++)
                 {
-                    var cachedShape = m_SensorShapes[i];
-                    var sensorShape = sensors[i].GetObservationShape();
-                    Debug.Assert(cachedShape.Length == sensorShape.Length, "Sensor dimensions must match.");
-                    for (var j = 0; j < Mathf.Min(cachedShape.Length, sensorShape.Length); j++)
-                    {
-                        Debug.Assert(cachedShape[j] == sensorShape[j], "Sensor sizes must match.");
-                    }
+                    var cachedSpec = m_SensorShapes[i];
```
Comment (chriselion) [inline com.unity.ml-agents/Runtime/Sensors/SensorShapeValidator.cs:37]: This got a little simpler (and probably faster)

File: com.unity.ml-agents/Tests/Editor/InplaceArrayTests.cs [hunk 12]
```diff
@@ -0,0 +1,193 @@
+using System;
+using System.Collections;
+using Boo.Lang.Runtime;
```
Comment (chriselion) [inline com.unity.ml-agents/Tests/Editor/InplaceArrayTests.cs:3]: Gah, fat-fingered something that made Rider add this. It's breaking a lot of yamato tests.

File: DevProject/ProjectSettings/Packages/com.unity.testtools.codecoverage/Settings.json (hunk 1 above)
Comment (surfnerd) [reply DevProject/ProjectSettings/Packages/com.unity.testtools.codecoverage/Settings.json:5]: yeah, I think it's just for local runs of code coverage

File: com.unity.ml-agents.extensions/Runtime/Sensors/GridSensor.cs (hunk 3 above)
Comment (surfnerd) [reply com.unity.ml-agents.extensions/Runtime/Sensors/GridSensor.cs:923]: I would lean toward yes

Comment (chriselion) [reply com.unity.ml-agents.extensions/Runtime/Sensors/GridSensor.cs:923]: OK, I'm going to punt that to a separate PR.

File: com.unity.ml-agents/Runtime/Sensors/ObservationSpec.cs (hunk 9 above)
Comment (dongruoping) [reply com.unity.ml-agents/Runtime/Sensors/ObservationSpec.cs:17]: I think we want to expose the ObservationSpec() constructor, and with that it's flexible enough for users to customize their own and not necessary to expose the fields

We check in the constructor that the length of shape and dimensionProperties match. Users can potentially set shape and dimensionProperties with different length if we expose the fields.

Comment (chriselion) [reply com.unity.ml-agents/Runtime/Sensors/ObservationSpec.cs:17]: OK, you convinced me. I'll make the fields internal and add public get properties.

File: com.unity.ml-agents/Runtime/Sensors/ObservationSpec.cs (hunk 10 above)
Comment (chriselion) [reply com.unity.ml-agents/Runtime/Sensors/ObservationSpec.cs:38]: Renamed to Rank

File: com.unity.ml-agents/Runtime/Sensors/ObservationSpec.cs (hunk 9 above)
Comment (chriselion) [reply com.unity.ml-agents/Runtime/Sensors/ObservationSpec.cs]: Updated.

File: Project/Assets/ML-Agents/Examples/Soccer/TFModels/SoccerTwos.onnx.meta [hunk 13]
```diff
@@ -1,9 +1,10 @@
 fileFormatVersion: 2
```
Comment (surfnerd) [inline Project/Assets/ML-Agents/Examples/Soccer/TFModels/SoccerTwos.onnx.meta:1]: thank you, i saw this before and it was annoying me

File: com.unity.ml-agents/Runtime/Sensors/ObservationSpec.cs [hunk 14]
```diff
@@ -0,0 +1,136 @@
...
+        {
+            get { return Shape.Length; }
+        }
+
+        /// <summary>
+        /// Construct an ObservationSpec for 1-D observations of the requested length.
+        /// </summary>
+        /// <param name="length"></param>
+        /// <returns></returns>
+        public static ObservationSpec Vector(int length)
```
Comment (vincentpierre) [inline com.unity.ml-agents/Runtime/Sensors/ObservationSpec.cs:63]: For Vector and Visual, add ObservationType to the static constructors.

File: com.unity.ml-agents/Runtime/InplaceArray.cs [hunk 15]
```diff
@@ -0,0 +1,250 @@
...
+{
+    /// <summary>
+    /// An array-like object that stores up to four elements.
+    /// This is a value type that does not allocate any additional memory.
+    /// </summary>
+    /// <remarks>
+    /// This does not implement any interfaces such as IList, in order to avoid any accidental boxing allocations.
+    /// </remarks>
+    /// <typeparam name="T"></typeparam>
+    public struct InplaceArray<T> where T : struct
```
Comment (vincentpierre) [inline com.unity.ml-agents/Runtime/InplaceArray.cs:14]: I would add a 
```
public struct ObservationShape : InplaceArray<int> { }
```
And same for DimensionProperties. Thoughts? Maybe overkill

**Review by vincentpierre — APPROVED** [review]
I like what I see

File: Project/Assets/ML-Agents/Examples/Soccer/TFModels/SoccerTwos.onnx.meta (hunk 13 above)
Comment (chriselion) [reply Project/Assets/ML-Agents/Examples/Soccer/TFModels/SoccerTwos.onnx.meta:1]: Oh, it was accidental (but also annoying me).

File: com.unity.ml-agents/Runtime/Sensors/ObservationSpec.cs (hunk 14 above)
Comment (chriselion) [reply com.unity.ml-agents/Runtime/Sensors/ObservationSpec.cs:63]: but make them default to Default? Sure, that sounds reasonable.

File: com.unity.ml-agents/Runtime/Analytics/Events.cs [hunk 16]
```diff
@@ -101,13 +101,14 @@ internal struct EventObservationSpec
...
-            var dimProps = (sensor as IDimensionPropertiesSensor)?.GetDimensionProperties();
+            var obsSpec = sensor.GetObservationSpec();
+            var shape = obsSpec.Shape;
+            var dimProps = obsSpec.DimensionProperties;
             var dimInfos = new EventObservationDimensionInfo[shape.Length];
             for (var i = 0; i < shape.Length; i++)
             {
                 dimInfos[i].Size = shape[i];
-                dimInfos[i].Flags = dimProps != null ? (int)dimProps[i] : 0;
+                dimInfos[i].Flags = (int)dimProps[i];
```
Comment (surfnerd) [inline com.unity.ml-agents/Runtime/Analytics/Events.cs:111]: dats nice

File: com.unity.ml-agents/Runtime/InplaceArray.cs (hunk 15 above)
Comment (chriselion) [reply com.unity.ml-agents/Runtime/InplaceArray.cs:14]: I think the constructors wouldn't work without adding them to ObservationShape too (plus DimensionProperties), which I'd rather not do.

And `using ObservationShape = InplaceArray<int>;` is only scoped for the file it's declared in, so wouldn't help any calling code.

Comment (dongruoping) [inline com.unity.ml-agents/Runtime/InplaceArray.cs:14]: Is there any plans of adding methods for converting between this and barracuda tensorshape?

**Review by dongruoping — APPROVED** [review]

File: com.unity.ml-agents/Runtime/Sensors/ObservationSpec.cs (hunk 14 above)
Comment (chriselion) [reply com.unity.ml-agents/Runtime/Sensors/ObservationSpec.cs:63]: @vincentpierre Any reason to not add that to VariableLength also?

File: com.unity.ml-agents/Runtime/InplaceArray.cs (hunk 15 above)
Comment (chriselion) [reply com.unity.ml-agents/Runtime/InplaceArray.cs]: Maybe we could do it as an extension method somewhere. Where would it be used?

File: com.unity.ml-agents/Runtime/Sensors/ObservationSpec.cs (hunk 14 above)
Comment (vincentpierre) [reply com.unity.ml-agents/Runtime/Sensors/ObservationSpec.cs]: The goal conditioning trainer will not work with VariableLength obs. We can add it though, but the trainer will raise an error or a warning when this happens.

File: com.unity.ml-agents/Runtime/InplaceArray.cs (hunk 15 above)
Comment (chriselion) [reply com.unity.ml-agents/Runtime/InplaceArray.cs]: Let's save this for another PR - don't think it'll affect the public API

File: com.unity.ml-agents/Runtime/Sensors/ObservationSpec.cs (hunk 14 above)
Comment (chriselion) [reply com.unity.ml-agents/Runtime/Sensors/ObservationSpec.cs]: Good call, removed the variable length.

**Pull request merged**